### PR TITLE
Implemented Integrity::from_hash(), inverse of Integrity::to_hash()

### LINKED ISSUE #10: need ssi::from_hex()
The garbageCollect() call this function.
```
const integrity = ssri.fromHex(digest, algo)
```
to get integrity from a content path, which contains algorithm and hash hex string.

Many thanks.

### PR DESCRIPTION
**Pull request by nouritsu**
Includes doc-comment, doc-test and `from_hash()` test in `mod tests`.
Fixes #10

### CODE REVIEW

File: src/integrity.rs [hunk 1]
```diff
@@ -138,6 +138,23 @@ impl Integrity {
...
+    /// # Example
+    ///```
+    /// use ssri::{Integrity, Algorithm};
+    ///
+    /// let expected = Integrity::from(b"hello");
+    /// let hex = String::from("2cf24dba5fb0a30e26e83b2ac5b9e29e1b161e5c1fa7425e73043362938b9824");
+    /// assert_eq!(Integrity::from_hex(hex, Algorithm::Sha256), expected);
+    ///```
+    pub fn from_hex<B: AsRef<[u8]>>(hex: B, algorithm: Algorithm) -> Integrity {
+        let digest = base64::prelude::BASE64_STANDARD.encode(hex::decode(hex).unwrap());
```
Comment (zkat) [inline src/integrity.rs]: Unlike `to_hex`, I think this should not `.unwrap()`. You should return a `Result` here. You'll have to add a variant to https://github.com/zkat/ssri-rs/blob/latest/src/errors.rs that wraps [`hex::FromHexError`](https://docs.rs/hex/latest/hex/enum.FromHexError.html)

**Comment by nouritsu** [comment]
Implemented returning `Result<Integrity, Error>`

**Comment by nouritsu** [comment]
@zkat I think this is ready to merge.

File: src/errors.rs [hunk 2]
```diff
@@ -14,4 +14,8 @@ pub enum Error {
     #[error("Integrity check failed.\n\tWanted: {0}\n\tActual: {1}")]
     #[diagnostic(code(ssri::integrity_check_error), url(docsrs))]
     IntegrityCheckError(Integrity, Integrity),
+    /// Error Decoding Hex Data
+    #[error("Failed decode hexadecimal data, reason: {0}")]
```
Comment (zkat) [inline src/errors.rs:18]:
```suggestion
    #[error(transparent)]
```

File: src/errors.rs [hunk 3]
```diff
@@ -14,4 +14,8 @@ pub enum Error {
     #[error("Integrity check failed.\n\tWanted: {0}\n\tActual: {1}")]
     #[diagnostic(code(ssri::integrity_check_error), url(docsrs))]
     IntegrityCheckError(Integrity, Integrity),
+    /// Error Decoding Hex Data
+    #[error("Failed decode hexadecimal data, reason: {0}")]
+    #[diagnostic(code(ssri::hex_decode_error), url(docsrs))]
+    HexDecodeError(String),
```
Comment (zkat) [inline src/errors.rs:20]:
```suggestion
    HexDecodeError(hex::FromHexError),
```

File: src/integrity.rs [hunk 4]
```diff
@@ -138,6 +138,27 @@ impl Integrity {
...
+    /// use ssri::{Integrity, Algorithm};
+    ///
+    /// let expected = Integrity::from(b"hello");
+    /// let hex = String::from("2cf24dba5fb0a30e26e83b2ac5b9e29e1b161e5c1fa7425e73043362938b9824");
+    /// assert_eq!(Integrity::from_hex(hex, Algorithm::Sha256).unwrap(), expected);
+    ///```
+    pub fn from_hex<B: AsRef<[u8]>>(hex: B, algorithm: Algorithm) -> Result<Integrity, Error> {
+        let b16 = match hex::decode(hex) {
+            Ok(data) => data,
+            Err(e) => return Err(Error::HexDecodeError(e.to_string())),
```
Comment (zkat) [inline src/integrity.rs]:
```suggestion
            Err(e) => return Err(Error::HexDecodeError(e)),
```

**Review by zkat — COMMENTED** [review]
Should use the actual error from `hex`

**Comment by nouritsu** [comment]
I'll make the changes and commit. My bad I'm extremely new to error handling 😅

**Comment by nouritsu** [comment]
Unfortunately hex::FromError does not implement Eq. Which leads to the following error -
image
Issue for implementing Eq seems to be open [here](https://github.com/KokaKiwi/rust-hex/issues/76#issue-1493158260).

**Comment by zkat** [comment]
SIGH. Yeah, it can't be helped. Just keep it as `String`, I guess :\

**Review by zkat — APPROVED** [review]

**Comment by zkat** [comment]
This has been released as part of `ssri@9.2.0` on crates.io. Thanks for the patch!

**Comment by nouritsu** [comment]
I've created a [PR](https://github.com/KokaKiwi/rust-hex/pull/80) to add `Eq` to `FromHexError`. If it gets merged, I will make another PR to this repository to make the error type wrap `FromHexError` rather than a `String`

**Comment by nouritsu** [comment]
@zkat an [alternate PR](https://github.com/KokaKiwi/rust-hex/pull/94) was merged since. I would like to make a patch that changes the error type.

**Pull request merged**